### ISSUE TITLE
optics: generalization? alternative?

### ISSUE THREAD

**Issue by aplavin** [cscherrer/NestedTuples.jl#23]
Noticed this package, and see that it already uses `Accessors` a little bit. I think more concepts from `NestedTuples` can naturally be represented within the optics paradigm. For example, flattening:
```julia
julia> using AccessorsExtra  # performant recursive optics not (yet) in Accessors proper

julia> nt = (x=1, y=(a=2, b=3))

julia> getall(nt, RecursiveOfType(Number))
(1, 2, 3)
```
recursive map:
```julia
julia> modify(x->x+1, nt, RecursiveOfType(Number))
(x = 2, y = (a = 3, b = 4))
```
leaf setter:
```julia
julia> setall(nt, RecursiveOfType(Number), (10, 20, 30))
(x = 10, y = (a = 20, b = 30))
```

An important feature of optics is that they straightforwardly compose, and have a shared interface. They are also more general - support arbitrary data structures, not just tuples.

I wonder what your thoughts are on these, and how optics compare to nestedtuples approach.

**Comment by cscherrer** [comment]
Hi @aplavin , thanks for the issue. I don't think I had seen AccessorsExtra. If it covers the functionality of this package with at least the performance (especially for deeply-nested named tuples, which can be very fiddly) we could consider deprecating this package in favor of that.

**Comment by aplavin** [comment]
Yeah, `AccessorsExtra` is not that old in the grand scheme of things :) Some tools such as `RecursiveOfType` are likely to be upstreamed into `Accessors` proper in a reasonable future.

I didn't intend to say that the current Accessors+ subsume _all_ functionality of `NestedTuples`, just that there is a significant overlap.
For example, stuff like `@with` and `randnt` don't have an obvious counterpart in the optics world. Other functionality may require a bit more boilerplate in the general optics framework, compared to this package that is specific to namedtuples.

Optics are great for composability, and maybe flatten/rmap/leaf_setter can be made more general and composable by using/exposing optics. For now that's just a suggestion to look in this direction though :)

**Comment by cscherrer** [comment]
Ok, thanks for pointing this out. I agree it's worth digging into to better understand how everything ties together.

If you do happen to see cases before I do where functionality is subsumed with no performance hit, please let me know so we can change the implementation to work in terms of the more general approach.
